doc/man3/ASN1_item_sign.pod

```diff
@@ -0,0 +1,223 @@
+=pod
+
+=head1 NAME
+
+ASN1_item_sign, ASN1_item_sign_ex, ASN1_item_sign_ctx,
+ASN1_item_verify, ASN1_item_verify_ex, ASN1_item_verify_ctx -
+ASN1 sign and verify
+
+=head1 SYNOPSIS
+
+ #include <openssl/x509.h>
+
+ int ASN1_item_sign_ex(const ASN1_ITEM *it, X509_ALGOR *algor1,
+                       X509_ALGOR *algor2, ASN1_BIT_STRING *signature,
+                       const void *data, const ASN1_OCTET_STRING *id,
+                       EVP_PKEY *pkey, const EVP_MD *md, OSSL_LIB_CTX *libctx,
+                       const char *propq);
+
+ int ASN1_item_sign(const ASN1_ITEM *it, X509_ALGOR *algor1, X509_ALGOR *algor2,
+                    ASN1_BIT_STRING *signature, const void *data,
+                    EVP_PKEY *pkey, const EVP_MD *md);
+
+ int ASN1_item_sign_ctx(const ASN1_ITEM *it, X509_ALGOR *algor1,
+                        X509_ALGOR *algor2, ASN1_BIT_STRING *signature,
+                        const void *data, EVP_MD_CTX *ctx);
+
+ int ASN1_item_verify_ex(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                         const ASN1_BIT_STRING *signature, const void *data,
+                         const ASN1_OCTET_STRING *id, EVP_PKEY *pkey,
+                         OSSL_LIB_CTX *libctx, const char *propq);
+
+ int ASN1_item_verify(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                      const ASN1_BIT_STRING *signature, const void *data,
+                      EVP_PKEY *pkey);
+
+ int ASN1_item_verify_ctx(const ASN1_ITEM *it, const X509_ALGOR *alg,
+                          const ASN1_BIT_STRING *signature, const void *data,
+                          EVP_MD_CTX *ctx);
+
+=head1 DESCRIPTION
+
+ASN1_item_sign_ex() is used to sign arbitrary ASN1 data using a data object
+I<data>, the ASN.1 structure I<it>, algorithm identifiers I<algor1> and I<algor2>,
+private key I<pkey> and message digest I<md>.
+The data that is signed is formed by taking the data object in I<data> and
+converting it to der format using the ASN.1 structure I<it>.
+Valid values that can be used by the ASN.1 structure I<it> are
+ASN1_ITEM_rptr(X509_CINF), ASN1_ITEM_rptr(X509_REQ_INFO) and
+ASN1_ITEM_rptr(X509_CRL_INFO).
+The B<OSSL_LIB_CTX> specified in I<libctx> and the property query string
+specified in I<props> are used when searching for algorithms in providers.
+The generated signature is set into I<signature>.
+The optional parameter I<id> can be NULL, but can be set for special key types.
+See EVP_PKEY_CTX_set1_id() for further info. The output parameters <algor1> and
+I<algor2> are set to the AlgorithmIdentifier that corresponds to the
```

Comment:
From this it is not clear what is the difference between algor1 and algor2 or why do we even have two parameters.

Reply:
Any suggestions?..
As you can see it is not easy to describe..
It allows you to optionally put the algorithm into an info object you are going to sign as well as a cached copy in a higher level object that contains the info object. From the point of view of this function it just allows you to set it twice...

@kroeckx said this above..
The to be signed data, and the struct with the signature can both have copy of the X509_ALGOR. The ASN1_item_sign() function will write the correct X509_ALGOR to the structs based on the algorithms and parameters that have been set up. If one of the algors points to the X509_ALGOR of the TBS data, that X509_ALGOR will first be written before the signature is generated.

Reply:
Can you just use what @kroeckx said?

Reply:
updated with a variant of the above.